components/style/properties/longhand/box.mako.rs

```diff
@@ -1144,39 +1145,453 @@ ${helpers.predefined_type("scroll-snap-coordinate",
+        TranslateZ(specified::Length),
+        Translate3D(specified::LengthOrPercentage,
+                    specified::LengthOrPercentage,
+                    specified::Length),
+    }
+
+    impl fmt::Display for SpecifiedTranslate {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            match *self {
+                SpecifiedTranslate::Translate(ref tx, ref ty) => {
+                    try!(f.write_str("translate("));
+                        try!(tx.to_css(f));
+                        try!(f.write_str(", "));
+                        try!(ty.to_css(f));
+                    f.write_str(")")
+                }
+                SpecifiedTranslate::TranslateX(ref tx) => {
+                    try!(f.write_str("translateX("));
+                        try!(tx.to_css(f));
+                    f.write_str(")")
+                }
+                SpecifiedTranslate::TranslateY(ref ty) => {
+                    try!(f.write_str("translateY("));
+                        try!(ty.to_css(f));
+                    f.write_str(")")
+                }
+                SpecifiedTranslate::TranslateZ(ref tz) => {
+                    try!(f.write_str("translateZ("));
+                        try!(tz.to_css(f));
+                    f.write_str(")")
+                }
+                SpecifiedTranslate::Translate3D(ref tx, ref ty, ref tz) => {
+                    try!(f.write_str("translate3d("));
+                        try!(tx.to_css(f));
+                        try!(f.write_str(", "));
+                        try!(ty.to_css(f));
+                        try!(f.write_str(", "));
+                        try!(tz.to_css(f));
+                    f.write_str(")")
+                }
+            }
+        }
+    }
+
+    impl ToCss for SpecifiedTranslate {
+        fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+            write!(dest, "{}", self)
+        }
+    }
+
+    impl HasViewportPercentage for SpecifiedTranslate {
+        fn has_viewport_percentage(&self) -> bool {
+            match *self {
+                SpecifiedTranslate::Translate(ref tx, ref ty) =>
+                    tx.has_viewport_percentage() || ty.has_viewport_percentage(),
+                SpecifiedTranslate::TranslateX(ref tx) =>
+                    tx.has_viewport_percentage(),
+                SpecifiedTranslate::TranslateY(ref ty) =>
+                    ty.has_viewport_percentage(),
+                SpecifiedTranslate::TranslateZ(ref tz) =>
+                    tz.has_viewport_percentage(),
+                SpecifiedTranslate::Translate3D(ref tx, ref ty, ref tz) =>
+                    tx.has_viewport_percentage() || ty.has_viewport_percentage() || tz.has_viewport_percentage(),
+            }
+        }
+    }
+
+    #[derive(Copy, Clone, Debug, PartialEq)]
+    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+    pub enum SpecifiedScale {
+        Scale(CSSFloat, CSSFloat),
```

Comment:
SpecifiedScale will print `scale(1.25, 1.25)` if we write something like this: `scale(1.25)`. It should preserve the one value form. We can make the second float optional and print if it present